NETOBSERV-1064: TCP drop & DNS tracking

jpinsonneau opened the pull request:
Adding `packet loss` query options as:
- Dropped: show records fully dropped packets
- Contains drops: show records that contains drops but may be sent
- Sent: show records that contains sent packets
- All: show all records

![image](https://user-images.githubusercontent.com/91894519/235191395-59f89bba-38b4-4bcb-94ec-bdb681ebc7e6.png)

Also add drops count in `Packets` & `Bytes` columns:
![image](https://user-images.githubusercontent.com/91894519/235190878-ae327fa5-2bbb-452b-b2c4-46d7a01f3af9.png)
Side panel detailed view:
![image](https://user-images.githubusercontent.com/91894519/235190967-1c30b013-4952-422c-8122-734c912355c2.png)

Full page:
![image](https://user-images.githubusercontent.com/91894519/235191087-fe407a7b-99c9-4f50-9864-d5423ad5c0d4.png)

Topology redlines:
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/5dc54bab-3582-464c-a52b-45bc91ed1268)

Overview new graphs:
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/e2d3dd9a-1ed8-4809-8394-3f5bd8025152)
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/b5652503-5203-4d04-9258-7460a1ace173)
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/bbcd522a-791a-4123-9f28-36401284cd43)

Adding `DNS request / response / latency`:
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/80a19748-6537-4937-987a-28457bf0fdfc)

Summary panel:
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/ca4f5925-6e9c-40b9-82de-49d36e5dab37)

The console plugin check for `features: ['tcpDrop', 'dnsTracking']` in its ConfigMap to show / hide the additionnal panels / columns.

ToDo: 
- [X] Dark theme
- [X] Topology lines color
- [X] ~Topology selection~ actually works fine on 4.13.4
- [X] TcpDropState graph
- [X] TcpDropCause graph
- [X] ~Exclude "0 = empty dropped" state / "not dropped yet" cause from queries~ Replaced by [full text](https://github.com/netobserv/flowlogs-pipeline/pull/429#discussion_r1219203830)

You can use `make start-standalone-mock` to try this PR without running related ones:
https://github.com/netobserv/netobserv-ebpf-agent/pull/115

eranra commented:
@jpinsonneau Looks good 
(1) can we agree to "remove" the entries when there is N/A --- I do not see a logic to showing red-colored N/A values 
(2) Looking into the UI screenshot - I see partially dropped packets in some entries --- is that correct? Do we drop parts of the data from a flow? This is interesting ( maybe this is due to sampling? and if so, this will cause a lot of questions ... is the data correct )

msherif1234 commented:
/ok-to-test

jpinsonneau commented on web/src/utils/tcp-drop.ts:
```diff
@@ -0,0 +1,296 @@
+/** TODO: remove this file, replaced by https://github.com/netobserv/flowlogs-pipeline/pull/429/files#r1219203830 */
```
I keep this file until we validate [full text](https://github.com/netobserv/flowlogs-pipeline/pull/429#discussion_r1219203830) approach

msherif1234 replied:
is it time now to remove this file ?

jpinsonneau replied:
https://github.com/netobserv/network-observability-console-plugin/pull/324/commits/398fa08a3a668788af70b1556cea6cce800cd5b9

jpinsonneau commented on web/src/utils/icmp.ts:
```diff
@@ -0,0 +1,75 @@
+/** TODO: remove this file, when replaced https://github.com/netobserv/flowlogs-pipeline/pull/429/files#r1219203830 */
```
Similar to tcp drop approach, we could also apply this on FLP side

jpinsonneau replied:
Not sure what's the plan here; let's wait for feedback. 

We'll do a followup anyways if it's the case

jotak replied:
not sure to follow that discussion :-) what are the options?
is it about who is responsible to interpret the values: flp or plugin ?

jpinsonneau replied:
The TCP State & Cause [are translated to string in FLP](https://github.com/netobserv/flowlogs-pipeline/pull/429#discussion_r1219203830) whereas it's not the case for ICMP values.

I wonder if we should do it or not

jotak commented on pkg/loki/flow_query.go:
```diff
@@ -55,14 +56,39 @@ func NewFlowQueryBuilder(cfg *Config, start, end, limit string, reporter constan
 		}
 	}
+
+	extraLineFilters := []string{}
```
if necessary we can create a new filter type, if the existing ones don't work for this use case - but that would avoid having several fields for apparently the same purpose

jpinsonneau replied:
that's right; that should do the work: https://github.com/netobserv/network-observability-console-plugin/pull/324/commits/840530f8f51ece05c768351a61d24cbef597be8e

Thanks @jotak !

jotak commented on web/src/api/loki.ts:
```diff
@@ -81,6 +83,14 @@ export interface TopologyMetricPeer {
 }
 
+export type DroppedTopologyMetrics = {
+  value: number;
```
looking only at this type definition, it is unclear what is `value` .. also I can't see any reference to that field, is it unused?

jpinsonneau replied:
correct that's not used anymore
https://github.com/netobserv/network-observability-console-plugin/pull/324/commits/4d1d412011488b8cd012b7ee61fe5be676ccb32f

jotak commented on pkg/loki/flow_query.go:
```diff
@@ -55,14 +56,39 @@ func NewFlowQueryBuilder(cfg *Config, start, end, limit string, reporter constan
+		} else if packetLoss == constants.PacketLossSent {
+			// match 1+ packets sent
+			sb.WriteString("|~`")
+			sb.WriteString(`Packets":[1-9]*[,}]`)
```
bug spotted: I'm seeing dropped packet in this mode that shouldn't have any, and just figured out why: the regex `Packets":[1-9]*[,}]` matches something like `"TcpDropPackets":0,` 
I guess you should include the leading `"`

jotak replied:
Another issue: all these regexs won't match packets or dropped with digit 0 in the number, such as "10". They should be instead: `[1-9][0-9]*[,}]`

jotak commented on web/src/components/dropdowns/query-options-dropdown.tsx:
```diff
@@ -203,6 +230,51 @@ export const QueryOptionsPanel: React.FC<QueryOptionsDropdownProps> = ({
+        <Tooltip
+          content={t(
+            // eslint-disable-next-line max-len
+            'Records to show from dropped, sent or mix of both.'
```
I think it deserves a more detailed explanation, as it doesn't seem trivial to understand all the options.

For instance I didn't immediately understand what was the difference between Dropped and Contains drops, especially when it happens to show the exact same thing (which seems to be often the case in the flow table). So, maybe saying something like "Dropped" shows the flows that are 100% dropped whereas "Contains drops" shows the flows having at least one packet dropped.

Also, when reading "Packet loss" > "All" , it can be ambiguous as we may understand that it's going to show all drops. It's probably the title "Packet loss" that is misleading, maybe it should be "Drops filter" ? And options like:
- Fully dropped
- Contains drops
- No drops
- All (or any?)

cc @stleerh any take on this?

jotak replied:
cc @Amoghrd too

jpinsonneau replied:
Here we go https://github.com/netobserv/network-observability-console-plugin/pull/324/commits/e5edc3039eb5f407d1593ac20ccd1dfbeaaeae67

![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/76e34441-7709-4782-bedc-31a55d9a8ab6)

jpinsonneau commented:
Fully retested in OCP console with addressed feedback:
- updated descriptions
  - `without drops` (previously sent) filter now match on `"TcpDropPackets":0`
- force strict label check
- reuse `lineFilters`

![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/089e5315-d979-4b7a-ab6b-4095dc9470dd)
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/0956905c-0332-4f4e-b61e-22649b108866)
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/1c3740c5-2c05-4605-a59e-bcc1a50906af)
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/6714c364-0675-421f-ba32-176569cab778)

jotak commented on web/src/components/metrics/metrics-helper.tsx:
```diff
@@ -147,6 +147,9 @@ const getPeerName = (
   if (name) {
     return truncateParts(name, truncateLength);
   }
+  if (scope === 'droppedCause' || scope === 'droppedState') {
```
Could be for a follow-up PR: looks like these new values aren't really scopes, and since they need sometimes to be excluded, maybe we could compose types such as:

```ts
   type FlowScope = 'host' | ... // (usual flow scopes);
   type AggregateBy = FlowScope | 'droppedCause' | 'droppedState';
```

So we don't mix them too much which will avoid having to handle special cases like that

jpinsonneau replied:
here we go: https://github.com/netobserv/network-observability-console-plugin/pull/324/commits/826ac695d633e5ae975d9e6cfdf6b54615c17663

jotak commented on web/src/components/query-summary/summary-panel.tsx:
```diff
@@ -291,6 +293,41 @@ export const SummaryPanelContent: React.FC<{
+
+  const timeContent = () => {
+    const filteredFlows = flows || [];
+    const duration =
```
I guess a check is needed that `filteredFlows.length` is >0 (avoid div by 0)

jpinsonneau replied:
here we go: https://github.com/netobserv/network-observability-console-plugin/pull/324/commits/3cf1597b83ee456a7abeeebcbf04a6296a426993

jotak commented:
My main concern is for the whole DNS epic: I don't think it brings as much value as it could, because of the user workflow: having to find specific DNS request and response by playing with filters, before being eventually able to get the latency info, makes it only relevant for specific troubleshooting use cases and not for monitoring in a broader sense, and even then, the value added is limited as we can get the same info by looking at rq/rs flow timestamps.

I get that this is inherent to the data provided by the Agent, so maybe we need to discuss more about how we could enhance the experience (e.g. revisit how the agent provides the dns flows?)

This is not blocking this PR, but I think we should either find follow-ups to do as part of the DNS epic, or flag this epic as experimental, or something in that vein, if it's not fully fleshed and is required to be revisited.

Also, as follow-ups for DNS: should we do anything specific about the DNS flags? I see the field is added but nothing is done from it. E.g. filtering for DNS errors, showing errors in a specific way?

jpinsonneau commented on web/src/components/netflow-overview/netflow-overview.tsx:
```diff
@@ -72,9 +81,18 @@ export const NetflowOverview: React.FC<NetflowOverviewProps> = ({
   if (error) {
     return <LokiError title={t('Unable to get overview')} error={error} />;
   } else if (_.isEmpty(metrics) || !totalMetric) {
+    //TODO: manage each metrics loading state separately
```
https://issues.redhat.com/browse/NETOBSERV-1185

jotak commented:
/lgtm
thanks @jpinsonneau !

jpinsonneau commented:
> /lgtm thanks @jpinsonneau !

@jotak looks like rebasing makes me forget about a flag to expose DNS Id filter:
https://github.com/netobserv/network-observability-console-plugin/pull/324/commits/13c79091dc900a9808e98c86afeb6327d22c050b
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/b5d98e57-2f4b-47d4-bce5-154dae61316a)

Have you tested it ?

Amoghrd commented:
/ok-to-test

msherif1234 commented:
https://github.com/netobserv/network-observability-console-plugin/pull/356 included this PR

jpinsonneau commented:
Replaced by https://github.com/netobserv/network-observability-console-plugin/pull/356
